### PR TITLE
Validate code when scheduling uprades from registrar

### PR DESCRIPTION
**Pull request by Szegoo**
Currently, anyone can registrar a code that exceeds the code size limit when performing the upgrade from the registrar. This PR fixes that and adds a new test to cover this.

cc @bkchr @eskimor

### CODE REVIEW

File: polkadot/runtime/common/src/paras_registrar/mod.rs [hunk 1]
```diff
@@ -681,6 +681,16 @@ impl<T: Config> Pallet<T> {
 		debug_assert!(res2.is_ok());
 		T::OnSwap::on_swap(to_upgrade, to_downgrade);
 	}
+
+	/// Checks whether the provided code is valid.
+	fn validate_code(code: ValidationCode) -> DispatchResult {
+		let config = configuration::Pallet::<T>::config();
+
+		ensure!(code.0.len() > 0, Error::<T>::EmptyCode);
+		ensure!(code.0.len() <= config.max_code_size as usize, Error::<T>::CodeTooLarge);
```
Comment (bkchr) [inline polkadot/runtime/common/src/paras_registrar/mod.rs:690]: Looks like this should be done inside of `schedule_code_upgrade`?

**Review by bkchr — COMMENTED** [review]
I would move the check, otherwise it looks good.

**Review by seadanda — APPROVED** [review]
LGTM

File: polkadot/runtime/parachains/src/paras/mod.rs [hunk 2]
```diff
@@ -1230,6 +1234,10 @@ impl<T: Config> Pallet<T> {
 		// Check that we can schedule an upgrade at all.
 		ensure!(Self::can_upgrade_validation_code(id), Error::<T>::CannotUpgradeCode);
 		let config = configuration::Pallet::<T>::config();
+		// Validation code sanity checks:
+		ensure!(new_code.0.len() > 0, Error::<T>::EmptyCode);
```
Comment (s0me0ne-unkn0wn) [inline polkadot/runtime/parachains/src/paras/mod.rs:1238]: A compressed blob should contain at least 8 bytes of ZSTD signature and 1 byte of payload, thus 9 bytes.

If it's uncompressed, it should contain a 4-byte Wasm magic, a 4-byte version, and at least one section.

Thus, it's safe to assume the minimal possible length of valid code is 9 bytes.

But I'm also okay with keeping this a more general "just-non-zero" check.

**Review by s0me0ne-unkn0wn — APPROVED** [review]
Looks good! Left a nit suggestion, not a blocker.

File: polkadot/runtime/parachains/src/paras/mod.rs (hunk 2 above)
Comment (ggwpez) [reply polkadot/runtime/parachains/src/paras/mod.rs:1238]: Such assumptions should then probably be checked by the `ValidationCode` type itself, maybe with some `check_sanity` function.  
Accessing the `.0` here and making assumptions about it seems unclean to me.

Comment (s0me0ne-unkn0wn) [reply polkadot/runtime/parachains/src/paras/mod.rs:1238]: Fair enough! I meant just some stricter checks, but from the design point of view, you have totally valid point here.

**Review by eskimor — APPROVED** [review]

File: polkadot/runtime/parachains/src/paras/mod.rs [hunk 3]
```diff
@@ -1230,6 +1232,10 @@ impl<T: Config> Pallet<T> {
 		// Check that we can schedule an upgrade at all.
 		ensure!(Self::can_upgrade_validation_code(id), Error::<T>::CannotUpgradeCode);
 		let config = configuration::Pallet::<T>::config();
+		// Validation code sanity checks:
+		ensure!(new_code.check_sanity(), Error::<T>::InvalidCode);
+		ensure!(new_code.0.len() <= config.max_code_size as usize, Error::<T>::InvalidCode);
```
Comment (bkchr) [inline polkadot/runtime/parachains/src/paras/mod.rs:1237]: I still think that this should be moved to `schedule_code_upgrade`.

**Review by bkchr — APPROVED** [review]
Ty! Still some benchmarks are failing.

**Comment by Szegoo** [comment]
> Still some benchmarks are failing.

🙈Should be good now.

**Pull request merged**